### ISSUE TITLE
Consul KV and Vault for TLS certificate storage

### ISSUE THREAD

**Issue by Gufran** [mholt/caddy#1918]
This issue is a feature proposal, discussion thread, and progress tracker for said feature.

### 2. What are you trying to do?

I propose two new storage backends for SSL certificate:

 - [Consul][] KV
 - [Vault][]

currently Caddy supports only disk or in-memory storage, both of these are far from ideal in an
environment with hundreds of certificates being served by redundant machines, SaaS businesses are a good example here. It is easy to spam the certificate issuer if each of those machine has its own storage. There are ways to work around this limitation with shared volumes but at the very least it is inconvenient and prone to human errors or inconsistent configuration.

There are also community supported plugins available but they have their own development lifecycle, decisions around feature and behaviour, and usual obligations for the maintainers that come with an open source project.

Native support for these storages will tie previously mentioned loose ends and declare the standards.

### Proposed changes

`tls` block has an undocumented directive `storage` which lets one change the storage backend. This doc proposes a backwards compatible change to `storage` directive as follows:

```text
tls {
  storage [name] {
    key value
    ...
  }
}
```

The content of the block will be a map which will be passed down to the storage backend as it is.
`caddytls.Storage` interface will be updated to provide a new method:

``` go
ValidateConfig(map[string]string) error
```

which will be called as part of setup to validate the additional configuration provided to `storage` directive and can raise an error if the configuration is not understood by the storage driver.

`caddytls.StorageConstructor` will be updated to accept the configuration as second parameter:

```go
type StorageConstructor func(caURL *url.URL, config map[string]string) (Storage, error)
```

These changes together will provide scaffolding necessary for more complex storage backends than
disk or in-memory.

### Consul KV
Consul KV storage backend is the simpler one. Possible configuration options are:

| Attribute | Possible Value | Required | Description |
|-----------|----------------|----------|-------------|
| Host | valid URL | yes | Consul Host URL including the protocol and port |
| Prefix | any string | yes | KV path prefix under which the certificates will be stored |
| Token | any string | no | Consul ACL token with permissions to read and write under `prefix` |
| Datacenter | any string | no | Consul datacenter |
| Auth | tuple (username, password) | no | basic auth credentials for Consul client |
| TLS  | tuple (certificate, key) or `trust` or `off` | no | TLS configuration. On by default |
| AES Key | 32 characters long string | no | used to encrypt the contents. contents will be stored in plain text if not provided |

All of these values can also be sourced from environment if set under keyspace `CADDY_TLS_CONSUL_`

### Vault
Vault backend is a slight bit more complex than Consul KV. Possible configuration options are:

| Attribute | Possible Value | Required | Description |
|-----------|----------------|----------|-------------|
| Host | valid URL | yes | Vault Host URL including the protocol and port |
| Path | any string | yes | Vault backend path under which the certificates will be stored |
| Auth | tuple (method, argument) | yes | Argument can be omitted if AWS or GCP auth is used |
| TLS  | tuple (certificate, key) or `trust` or `off` | no | TLS configuration. On by default |
| Read only | true or false | no | False by default, set to true if backend does not support writing to it (e.g. PKI) |

All of these values can also be sourced from environment if set under keyspace `CADDY_TLS_VAULT_`

#### Problems

Following problems with Vault backend are not solved and any idea would be appreciated.

 - No shared locking mechanism
 - If token auth is used and Caddy fails to renew the token there is no reasonable course of actions to prevent downtime

If this is something that can add value for Caddy community I am willing to work on it.
If, for any reason, this addition cannot be accepted please feel free to close the issue otherwise I'll appreciate if a maintainer can assign it to me.

Lets discuss.

[Consul]: https://www.consul.io/
[Vault]: https://www.vaultproject.io/

**Comment by krishamoud** [comment]
I recently wrote a [consul-loader](https://github.com/krishamoud/caddy-consul) plugin for caddy.  I would be interested in helping.

**Comment by mholt** [comment]
Hi, thanks for writing this up; nice proposal.

One thing that I want to finish upgrading in Caddy is its storage API. Right now it interacts with the file system directly but it would be good to abstract that away for any of Caddy's storage needs to enable this. It's partway done, but not all the way. It would be nice if it worked for more than just caddytls.

As for the Caddyfile syntax: nested blocks are not supported, so within the `tls` directive, the storage configuration will need to be flattened. That, or use a different directive.

**Comment by Gufran** [comment]
> Right now it interacts with the file system directly but it would be good to abstract that away for any of Caddy's storage needs to enable this. It's partway done, but not all the way. It would be nice if it worked for more than just caddytls.

@mholt I'll be happy to work on that, but I'm not sure if I understand the problem completely.
Do you mean that any part of Caddy which require storage should use one universal storage abstraction?

**Comment by mholt** [comment]
Right, I need to go into more detail. Caddy uses disk storage for a few things: right now, mainly managed certificates (and their related metadata) and caching OCSP staples. Plus a few other things but the primary reason for abstracting away the file system is to run Caddy distributed, and that basically means anything in the `.caddy` folder should go through some sort of interface.

POSIX has a pretty standard interface that we could imitate, but I don't know if it would be difficult to make that work with non-file-systems, like KV stores or networky things. I like how generalizable the POSIX FS interface is, and I want to avoid a very specific interface with methods like "GetSite()" or "ListCertificateFiles()" and stuff like that. Do you think a POSIX-like interface could be satisfied by KV stores, Kubernetes setups, etc?

**Comment by Gufran** [comment]
@mholt In my opinion POSIX specification is unnecessarily complicated for our use case. I think an  interface as simple as

``` go
Put(key string, value []byte) error
Get(key string) ([]byte, error)
Delete(key string) error
```

is adequate.
It will work with any backend, disk backed or remote, doesn't matter.
I'd like to debate it though.

**Comment by mholt** [comment]
Unfortunately, Caddy requires a little more information, at least, the way things are implemented right now. For example, it scans the list of folders containing ACME CA account information to get the most recently used email address, to use that when negotiating certificates, if no email is provided. We could re-engineer this so that the most recent email used is stored as a KV itself, but keep that kind of thing in mind...

I agree that your interface is much simpler, but it's not sufficient even if we re-engineer a couple of functions that require iterating certain kinds of keys. Some operations will require obtaining a distributed exclusive lock, while others require _trying_ to obtain a lock but doing something else if a lock is already in use by another instance. The current prototype interface implemented in caddytls/storage.go has `TryLock(name string) (Waiter, error)` and `Unlock(name string) error` but these don't generalize very well.

Then when a lock is released, all waiters need to be notified...

**Comment by Gufran** [comment]
> it scans the list of folders containing ACME CA account information to get the most recently used email address, to use that when negotiating certificates, if no email is provided

@mholt This is a driver specific detail, consumer of the interface shouldn't know about it. so lets add another method to the interface and lets say ACME wraps the storage interface and exposes the storage operations supported by ACME.

``` go
type Storage interface {
  Put(key string, value []byte) error
  Get(key string) ([]byte, error)
  Delete(key string) error
  List(prefix string, recurse bool) (map[string][]byte, error)
}

type ACME struct {
  storge Storage
  // ...
}

func (a *ACME) GetEmailAddr() (string, error) {}
func (a *ACME) ListAllCertificates() ([]Cert, error) {}

```

and same thing for other packages.
Like David Wheeler said — _All problems in computer science can be solved by another level of indirection_. we wrap `Storage` in a more specific interface or struct as needed.

Distributed locks are not supported by all storage backends (vault!) so I propose that the the lock mechanism be made into a separate abstraction of its own. This will also allow us to use different storage backend and lock provider (disk and consul, for example). I think an interface like

``` go
type Locker interface {
  // Block until the lock is acquired or error out
  Lock(name string) (Waiter, error)

  // Return immediately with or without acquiring the lock, or error out.
  // Waiter will yield immediately If the lock couldn't be acquired.
  TryLock(name string) (Waiter, error)

  Unlock(name string) error
}
```

Having said all of this, I realise it is a bit of work and refactoring but it will greatly improve the underlying design.

**Comment by mholt** [comment]
I like where this is going, with the understanding that distributed locking is required no matter what the storage provider backend, so *something* has to implement it.

So how do you feel about this? Is this something you're willing to help undertake? :D

**Comment by Gufran** [comment]
@mholt sure. I'll be glad to work on it. let me hack something together this weekend and put it up for discussion.
On a side note, would you entertain the possibility to implement first class distributed locking support in Caddy? I have been thinking about it and I believe it shouldn't be hard using [go-discover](https://github.com/hashicorp/go-discover) and [memberlist](https://github.com/hashicorp/memberlist)

**Comment by mholt** [comment]
Sounds good to me. I will be happy to review!

> On a side note, would you entertain the possibility to implement first class distributed locking support in Caddy? I have been thinking about it and I believe it shouldn't be hard using go-discover and memberlist

That's interesting; I feel like this would probably be a separate implementation? Also, is the discovery mechanism the same as first-class distributed locking support? Or are you saying that you'd want first class service discovery beneath distributed locking.

**Comment by Gufran** [comment]
This will be a separate implementation, yes.

> is the discovery mechanism the same as first-class distributed locking support? Or are you saying that you'd want first class service discovery beneath distributed locking.
No. Discovery mechanism will only find other Caddy nodes that are going to share the storage.

Node discovery is only a bit of sugar to make configuration and deployment easier for people.
I'll quickly jot down what's on my mind, please excuse any oversight.

`go-discover` will discover other Caddy nodes and pass the information to `memberlist`. Cluster membership and failure detection is managed by `memberlist` package.
We trade consistency and become a leaderless (AP) cluster. Consistency does not matter as much as availability. In worst case we will end up provisioning more than one certificate but all of them will be valid, the storage will retain the last written certificate.

A node will broadcast the message (lock) when it need the lock, the cluster will vote in favor (grant) or against (deny) the node.
Since we are trading off consistency, a lock is granted if:

1. Whole cluster granted the lock
1. Majority granted the lock, minority is in other partition
1. Minority granted the lock, majority is in other partition
1. Majority granted the lock, minority denied

Lock is denied if:

1. Whole cluster denied the lock
1. Majority denied the lock, minority is in other partition
1. Majority denied the lock, minority granted
1. Minority denied the lock, majority is in other partition

Lock are granted with timeout and cannot be renewed. A node holding an expired lock will need to drop it ask for it again.
If a lock is denied then the node can choose to wait in queue for lock or continue with this knowledge.
Lock is released by broadcasting the message (release).
A node which has voted in favor of a lock must deny any subsequent request for the same lock until the lock is released or times out.

With abstraction around lock mechanism this will become another lock provider, people can still choose to utilize other lock providers, such as consul, if they wish.
This is all an idea and I probably missed a huge bunch of corner cases but I can discuss it in detail if it look interesting.

**Comment by mholt** [comment]
Does the discovery need to be built-in? It seems pretty heavyweight, considering that I think the majority of users won't use it. And it sounds like something separate from just enabling pluggable locking/storage. Can we do the pluggable locking/storage first and separate?

**Comment by Gufran** [comment]
@mholt discovery is actually the lightest and most trivial part in this.
And sure, pluggable storage and locking first and separate. I tore apart the codebase this weekend but didn't much time to work on it, I'll think about discovery and distributed locking when I'm done writing the abstraction.

**Comment by j0hnsmith** [comment]
Not sure if useful but here's another consul implementation https://github.com/pteich/caddy-tlsconsul. I made a [Google CDS  implementation](https://github.com/j0hnsmith/caddy-tlsclouddatastore) that has some, rather naive iirc, locking.

**Comment by miekg** [comment]
go-discover seems to tie this to cloud providers. What about a non cloud use-case? Multiple caddy instances that "just" want to share the TLS certs?

**Comment by Gufran** [comment]
@miekg the option to provide caddy with the addresses of other nodes is always there, I brought up go-discover because it automates everything.

**Comment by Gufran** [comment]
I'm sorry about lack of communication here, it is a big piece and I only get to work on it over the weekend.

**Comment by MOZGIII** [comment]
Could you add AWS S3 as well?

**Comment by mholt** [comment]
Related: with PR #2015, Caddy now supports synchronizing renewals between multiple instances as long as they use the same certificates on disk. In other words, if they mount the `$CADDYPATH/acme` folder as part of the file system, they will not duplicate renewals and they will safely share the same storage. So, if that is helpful to you, the next release (0.10.11) would be a good time to upgrade.

**Comment by MOZGIII** [comment]
TLDR: synchronization via shared file system? Like local ext4 or NFSv4?

**Comment by mholt** [comment]
> TLDR: synchronization via shared file system? Like local ext4 or NFSv4?

Yes, the files must be mounted locally as part of the file system. But if the disk is shared (e.g. a NAS) multiple Caddy instances can safely share those certificates.

**Comment by francislavoie** [comment]
Since this issue was written, https://github.com/pteich/caddy-tlsconsul has been implemented. @Gufran does this cover your usecases?

**Comment by Gufran** [comment]
The use case no longer exist but I believe https://github.com/pteich/caddy-tlsconsul would've solved it.

**Comment by francislavoie** [comment]
I think we can close this then.
